xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -1113,24 +1149,47 @@ protected void handleShutdown() {
     cleanUpResourceTimers();
   }
 
-  @Nullable
-  @Override
-  Collection<String> getSubscribedResources(ResourceType type) {
+  private Map<String, ResourceSubscriber> getSubscribedResourcesMap(ResourceType type) {
     switch (type) {
       case LDS:
-        return ldsResourceSubscribers.isEmpty() ? null : ldsResourceSubscribers.keySet();
+        return ldsResourceSubscribers;
       case RDS:
-        return rdsResourceSubscribers.isEmpty() ? null : rdsResourceSubscribers.keySet();
+        return rdsResourceSubscribers;
       case CDS:
-        return cdsResourceSubscribers.isEmpty() ? null : cdsResourceSubscribers.keySet();
+        return cdsResourceSubscribers;
       case EDS:
-        return edsResourceSubscribers.isEmpty() ? null : edsResourceSubscribers.keySet();
+        return edsResourceSubscribers;
       case UNKNOWN:
       default:
         throw new AssertionError("Unknown resource type");
     }
   }
 
+  @Nullable
+  @Override
+  Collection<String> getSubscribedResources(ResourceType type) {
+    Map<String, ResourceSubscriber> resources = getSubscribedResourcesMap(type);
+    return resources.isEmpty() ? null : resources.keySet();
+  }
+
+  Map<String, ResourceMetadata> getSubscribedResourcesMetadata(ResourceType type) {
+    Map<String, ResourceMetadata> metadataMap = new HashMap<>();
+    for (Map.Entry<String, ResourceSubscriber> entry : getSubscribedResourcesMap(type).entrySet()) {
+      metadataMap.put(entry.getKey(), entry.getValue().metadata);
+    }
+    return metadataMap;
+  }
+
+  @Nullable
+  ResourceMetadata getSubscribedResourceMetadata(
+      ResourceType type, String resourceName) {
+    Map<String, ResourceSubscriber> resources = getSubscribedResourcesMap(type);
+    if (resources.isEmpty() || !resources.containsKey(resourceName)) {
```

Comment:
Isn't the first half redundant? The gain seems to be very little (if any)...

Reply:
Good 👀!